script/migrate_notification_system.rb

```diff
@@ -0,0 +1,126 @@
+# frozen_string_literal: true
+
+# Run it with rails runner, for example:
+# $ bundle exec rails runner migrate_notification_system.rb
+
+require 'progress_counter'
+
+module MigrateNotificationSystem
+
+  module_function
+
+  # Notifications that didn't exist in the old system
+  NEW_NOTIFICATIONS = %i[account_deleted application_plan_change_requested service_plan_change_requested
+                       invoices_to_review plan_downgraded credit_card_unstore_failed expired_credit_card_provider
+                       unsuccessfully_charged_invoice_provider unsuccessfully_charged_invoice_final_provider
+                       csv_data_export service_deleted].freeze
+
+  def scope
+    # Active users, belonging to an active provider and without notification preferences, except impersonation admins
+    User.active
+        .joins(:account)
+        .includes(:notification_preferences)
+        .where({ account: { provider: true, master: nil, state: 'approved' }, notification_preferences: { id: nil }})
+        .where.not( username: ThreeScale.config.impersonation_admin[:username])
+  end
+
+  def call
+    total = scope.count
+    puts "Users count: #{total}"
+
+    users = scope.find_each
+
+    each_with_progress_counter(users, total) do |user|
+      preferences = migrated_preferences(user)
+      user.notification_preferences.update!(preferences:)
+
+      # puts
+      # puts "----------------------"
+      # puts "Provider: #{user.account.id}, #{user.account.name}"
+      # puts "User: #{user.id}, #{user.username}"
+      # puts "Previous preferences: #{user.notification_preferences.preferences.symbolize_keys}"
+      # puts "Migrated preferences: #{preferences}"
+      # puts "Changed? #{user.notification_preferences.preferences.symbolize_keys != preferences}"
+      # puts
+    end
+    puts # To show the terminal prompt in the next line
+  end
+
+  def each_with_progress_counter(enumerable, count)
+    progress = ProgressCounter.new(count)
+    enumerable.each do |element|
+      progress.call
+      yield element
+    end
+  end
+
+  def migrated_preferences(user)
+    default_preferences = NotificationPreferences.default_preferences
+
+    migrated_preferences = provider_preferences(user.account)
+
+    default_preferences.merge(new_preferences_disabled).merge(migrated_preferences)
```

Comment:
might be slightly simpler to just disable all notifications and then merge the `migrated_preferences` instead of processing the lists of default and new preferences.

But I'm fine this to stay as is if it works because this is a one-off.

Reply:
> might be slightly simpler to just disable all notifications and then merge the `migrated_preferences` instead of processing the lists of default and new preferences.
> 
> But I'm fine this to stay as is if it works because this is a one-off.

You are probably right on this, but yes, since it takes only 3-4 minutes to complete and it's a one-off, I think it's fine to leave it.